codegen-server/python/src/main/kotlin/software/amazon/smithy/rust/codegen/server/python/smithy/generators/PythonServerEnumGenerator.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.server.python.smithy.generators
+
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.StringShape
+import software.amazon.smithy.model.traits.DocumentationTrait
+import software.amazon.smithy.model.traits.EnumTrait
+import software.amazon.smithy.rust.codegen.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.rustlang.asType
+import software.amazon.smithy.rust.codegen.rustlang.docs
+import software.amazon.smithy.rust.codegen.rustlang.rust
+import software.amazon.smithy.rust.codegen.rustlang.rustBlock
+import software.amazon.smithy.rust.codegen.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.server.python.smithy.PythonServerCargoDependency
+import software.amazon.smithy.rust.codegen.server.smithy.generators.ServerEnumGenerator
+import software.amazon.smithy.rust.codegen.smithy.CodegenMode
+import software.amazon.smithy.rust.codegen.smithy.RuntimeConfig
+import software.amazon.smithy.rust.codegen.smithy.RustSymbolProvider
+import software.amazon.smithy.rust.codegen.smithy.generators.docWithNote
+import software.amazon.smithy.rust.codegen.util.getTrait
+
+/**
+ * To share enums defined in Rust with Python, `pyo3` provides the `PyClass` trait.
+ * This class generates enums definitions, implements the `PyClass` trait and adds
+ * some utility functions like `__str__()` and `__repr__()`.
+ */
+class PythonServerEnumGenerator(
+    model: Model,
+    symbolProvider: RustSymbolProvider,
+    private val writer: RustWriter,
+    private val shape: StringShape,
+    enumTrait: EnumTrait,
+    private val runtimeConfig: RuntimeConfig,
+) : ServerEnumGenerator(model, symbolProvider, writer, shape, enumTrait, runtimeConfig) {
+    override var mode: CodegenMode = CodegenMode.Server
+
+    private val errorStruct = "${enumName}UnknownVariantError"
+    private val codegenScope = arrayOf(
+        "pyo3" to PythonServerCargoDependency.PyO3.asType(),
+        "SmithyPython" to PythonServerCargoDependency.SmithyHttpServerPython(runtimeConfig).asType()
+    )
+
+    override fun renderFromForStr() {
+        writer.rustTemplate(
+            """
+            ##[#{pyo3}::pyclass]
+            ##[derive(Debug, PartialEq, Eq, Hash)]
+            pub struct $errorStruct(String);
+            """,
+            *codegenScope
+        )
+        renderEnumImpl()
+    }
+
+    override fun renderEnum() {
```

Comment:
I personally don't like the `CodegenTarget` approach since it will become a lot of ifs when we add new languages. I'll try to refactor this and promote more code re-use if I can.

Reply:
I have reduced the code that is copied to basically zero using some smarter inheritance :D